bfv/encoder.go

```diff
@@ -108,85 +108,83 @@ func NewEncoder(params Parameters) Encoder {
 	pt = NewPlaintextLvl(ecd.params, level)
-	ecd.Encode(coeffs, pt)
+	ecd.Encode(values, pt)
 	return
 }
 
-// EncodeUint encodes an uint64 slice of size at most N on a pre-allocated plaintext.
-func (ecd *encoder) Encode(coeffs interface{}, pt *Plaintext) {
+// Encode encodes a slice of integers of type []uint64 or []int64 of size at most N into a pre-allocated plaintext.
+func (ecd *encoder) Encode(values interface{}, pt *Plaintext) {
 	ptRt := &PlaintextRingT{pt.Plaintext}
 
 	// Encodes the values in RingT
-	ecd.EncodeRingT(coeffs, ptRt)
+	ecd.EncodeRingT(values, ptRt)
 
 	// Scales by Q/t
 	ecd.ScaleUp(ptRt, pt)
 }
 
-// EncodeUintRingTNew encodes a slice of uint64 into a newly allocated PlaintextRingT.
-func (ecd *encoder) EncodeRingTNew(coeffs interface{}) (pt *PlaintextRingT) {
+// EncodeRingTNew encodes a slice of integers of type []uint64 or []int64 of size at most N into a newly allocated PlaintextRingT.
+func (ecd *encoder) EncodeRingTNew(values interface{}) (pt *PlaintextRingT) {
 	pt = NewPlaintextRingT(ecd.params)
-	ecd.EncodeRingT(coeffs, pt)
+	ecd.EncodeRingT(values, pt)
 	return
 }
 
-// EncodeUintRingT encodes a slice of uint64 into a pre-allocated PlaintextRingT.
-func (ecd *encoder) EncodeRingT(coeffs interface{}, pt *PlaintextRingT) {
+// EncodeRingT encodes a slice of integers of type []uint64 or []int64 of size at most N into a pre-allocated PlaintextRingT.
+// The input values are reduced modulo T before encoding.
+func (ecd *encoder) EncodeRingT(values interface{}, ptOut *PlaintextRingT) {
 
-	if len(pt.Value.Coeffs[0]) != len(ecd.indexMatrix) {
+	if len(ptOut.Value.Coeffs[0]) != len(ecd.indexMatrix) {
 		panic("invalid plaintext to receive encoding: number of coefficients does not match the ring degree")
 	}
 
-	switch coeffs := coeffs.(type) {
+	pt := ptOut.Value.Coeffs[0]
+	valLen := 0
+	switch values := values.(type) {
 	case []uint64:
-
-		if len(coeffs) > len(ecd.indexMatrix) {
-			panic("invalid input to encode: number of coefficients must be smaller or equal to the ring degree")
-		}
-
-		value := pt.Value.Coeffs[0]
-		for i := 0; i < len(coeffs); i++ {
-			value[ecd.indexMatrix[i]] = coeffs[i]
-		}
-		for i := len(coeffs); i < len(ecd.indexMatrix); i++ {
-			value[ecd.indexMatrix[i]] = 0
+		T := uint64(ecd.params.T())
+		for i := 0; i < len(values); i++ {
+			v := values[i]
+			if v < T {
+				pt[ecd.indexMatrix[i]] = v
+			} else {
+				pt[ecd.indexMatrix[i]] = v % T
+			}
 		}
+		valLen = len(values)
 	case []int64:
-
-		if len(coeffs) > len(ecd.indexMatrix) {
-			panic("invalid input to encode: number of coefficients must be smaller or equal to the ring degree")
-		}
-
-		value := pt.Value.Coeffs[0]
 		T := int64(ecd.params.T())
-		for i := 0; i < len(coeffs); i++ {
-			if coeffs[i] < 0 {
-				value[ecd.indexMatrix[i]] = uint64(T + coeffs[i])
+		for i := 0; i < len(values); i++ {
+			v := values[i] % T
+			if v >= 0 { // Go's modulo operation returns values in [-(t-1); t-1]
```

Comment:
Can simply just add T, as the NTT input can tolerate larger inputs than the modulus (up to 2^61 bits)

Reply:
Done. Simlpy adding T is not enough as the inputs might be smaller than -T.

Reply:
I meant do `v  := (values[i] % T) + T`

Reply:
This uses `%`, though. The current implementation matches the performance of master.